test/functional/p2p_seednode.py

```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+"""
+Test seednode interaction with the AddrMan
+"""
+import random
+
+from test_framework.test_framework import BitcoinTestFramework
+
+
+class P2PSeedNodes(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.disable_autoconnect = False
+
+    def test_no_seednode(self):
+        # Check that if no seednode is provided, the node proceeds as usual (without waiting)
+        with self.nodes[0].assert_debug_log(expected_msgs=[], unexpected_msgs=["Empty addrman, adding seednode", "Couldn't connect to peers from addrman after 10 seconds. Adding seednode"], timeout=10):
+            self.restart_node(0)
+
+    def test_seednode_empty_addrman(self):
+        seed_node = "0.0.0.1"
+        # Check that the seednode is added to m_addr_fetches on bootstrap on an empty addrman
+        with self.nodes[0].assert_debug_log(expected_msgs=[f"Empty addrman, adding seednode ({seed_node}) to addrfetch"], timeout=10):
+            self.restart_node(0, extra_args=[f'-seednode={seed_node}'])
+
+    def test_seednode_addrman_unreachable_peers(self):
+        seed_node = "0.0.0.2"
+        # Fill the addrman with unreachable nodes
+        for i in range(10):
+            ip = f"{random.randrange(128,169)}.{random.randrange(1,255)}.{random.randrange(1,255)}.{random.randrange(1,255)}"
+            port = 8333 + i
+            self.nodes[0].addpeeraddress(ip, port)
+
+        # Restart the node so seednode is processed again
+        with self.nodes[0].assert_debug_log(expected_msgs=[f"Couldn't connect to peers from addrman after 10 seconds. Adding seednode ({seed_node}) to addrfetch"], unexpected_msgs=["Empty addrman, adding seednode"], timeout=20):
```

Comment:
I've been trying to reproduce in my local setup but I'm unable. Did you manage to trigger this on your updated branch, or with this one?

Reply:
Without any additional changes. With `--enable-debug` and building with Clang.

I've retested on ecded92737c4173a30a43e612a21c6e1f24d1605, here is the combined log for `test/functional/p2p_seednode.py`:
[29605_timeout.combined_log.gz](https://github.com/bitcoin/bitcoin/files/15240761/29605_timeout.combined_log.gz)

It seems like I only have to bump 20 -> 23 seconds to prevent frequent timeouts. I've got IPv6 turned off but otherwise not much weird stuff going on AFAIK. Maybe something stands out if you compare my logs to yours.

Reply:
I haven't been able to reproduce the issue in my local setup, but I think we may be able to improve on this (plus reduce the waiting time) by mocking the node's time after restarting it

Reply:
Still failing on 1ed818a35d78a5bf6eaf0f72f3730041913b4859 for me.
[29605_timeout_1ed818a.combined_log.gz](https://github.com/bitcoin/bitcoin/files/15254533/29605_timeout_1ed818a.combined_log.gz)

### Excerpts

Startup:
```
test  2024-05-08T20:37:27.879000Z TestFramework (INFO): PRNG seed is: 2801220528353188963 
test  2024-05-08T20:37:27.879000Z TestFramework (DEBUG): Setting up network thread 
```
~8 seconds later we make the attempt to connect to the bogus IP:
```
node0 2024-05-08T20:37:35.701878Z (mocktime: 2024-05-08T20:37:46Z) [opencon] [net.cpp:413] [ConnectNode] [net] trying v1 connection 140.186.182.6:8333 lastseen=0.0hrs 
```
Within less than 5 seconds, the assert triggers:
```
test  2024-05-08T20:37:40.024000Z TestFramework (ERROR): Assertion failed 
                                   Traceback (most recent call last):
                                     File "/home/chris/Documents/Code/bitcoin-core/test/functional/test_framework/test_framework.py", line 132, in main
                                       self.run_test()
                                     File "/home/chris/Documents/Code/bitcoin-core/test/functional/p2p_seednode.py", line 50, in run_test
                                       self.test_seednode_addrman_unreachable_peers()
                                     File "/home/chris/Documents/Code/bitcoin-core/test/functional/p2p_seednode.py", line 43, in test_seednode_addrman_unreachable_peers
                                       with node.assert_debug_log(expected_msgs=[f"Couldn't connect to peers from addrman after {ADD_NEXT_SEEDNODE} seconds. Adding seednode ({seed_node}) to addrfetch"], unexpected_msgs=["Empty addrman, adding seednode"], timeout=ADD_NEXT_SEEDNODE):
```

Reply:
That's interesting. I guess it's still too close.

```
 node0 2024-05-08T20:37:35.200946Z [opencon] [net.cpp:2501] [ThreadOpenConnections] Fixed seeds are disabled
 node0 2024-05-08T20:37:35.254287Z [http] [httpserver.cpp:306] [http_request_cb] [http] Received a POST request for / from 127.0.0.1:60322
 node0 2024-05-08T20:37:35.254724Z [httpworker.0] [rpc/request.cpp:187] [parse] [rpc] ThreadRPCServer method=getblockcount user=__cookie__
 node0 2024-05-08T20:37:35.256529Z [http] [httpserver.cpp:306] [http_request_cb] [http] Received a POST request for / from 127.0.0.1:60322
 node0 2024-05-08T20:37:35.256960Z [httpworker.2] [rpc/request.cpp:187] [parse] [rpc] ThreadRPCServer method=getmempoolinfo user=__cookie__
 test  2024-05-08T20:37:35.258000Z TestFramework.node0 (DEBUG): RPC successfully started
 node0 2024-05-08T20:37:35.259402Z [http] [httpserver.cpp:306] [http_request_cb] [http] Received a POST request for / from 127.0.0.1:60322
 node0 2024-05-08T20:37:35.259821Z [httpworker.1] [rpc/request.cpp:187] [parse] [rpc] ThreadRPCServer method=setmocktime user=__cookie__
 node0 2024-05-08T20:37:35.701790Z (mocktime: 2024-05-08T20:37:46Z) [opencon] [addrman.cpp:782] [Select_] [addrman] Selected 140.186.182.6:8333 from new
 node0 2024-05-08T20:37:35.701878Z (mocktime: 2024-05-08T20:37:46Z) [opencon] [net.cpp:413] [ConnectNode] [net] trying v1 connection 140.186.182.6:8333 lastseen=0.0hrs
```

There's more than 10 secs (of mocked time) between we enter the thread and make that last attempt. Can you increase the wait to ~15 and see if it still fails?

Reply:
Adding 5 to the timeout fixes it for me. Modifying the line with the mocktime does not (I also tried removing that line which caused the test to fail again, proving its usefulness).
```
        with node.assert_debug_log(expected_msgs=[f"Couldn't connect to peers from addrman after {ADD_NEXT_SEEDNODE} seconds. Adding seednode ({seed_node}) to addrfetch"], unexpected_msgs=["Empty addrman, adding seednode"], timeout=ADD_NEXT_SEEDNODE + 5):
            self.restart_node(0, extra_args=[f'-seednode={seed_node}'])
            node.setmocktime(int(time.time()) + ADD_NEXT_SEEDNODE + 1)
```

Reply:
Yeah, the mock time seems to reduce the runtime by 5-10 secs. I'm guessing with 10 secs of wait it could be the case that the connection is still being tried, so the timeout hasn't been considered. I'll add 5 more.